 design/README.MD

```diff
@@ -0,0 +1,456 @@
+A Dafny tests can check multiple runtimes.
+But a simple test would only check the Dafny interfaces.
+
+For more details on how this works
+see the [polymorph development](#polymorph-development) section blow.
+
+This feature is used to test `smithy-dafny` on itself
+in the TestModels.
+This features is used by the Encryption SDK
+in the test vector runners
+to test interoperability between native runtimes.
+
+In this mode `smithy-dafny` will not produce Dafny types and interfaces.
+It will use the Dafny types and interfaces that already exist.
+It will also create a shim
+that will convert the Dafny to native
+and then call the native implementation.
+
+```mermaid
+---
+title: Wrapped Library
+---
+flowchart RL
+
+subgraph Wrapping[Wrapping Library]
+    direction RL
+    WrappingInterface[Generated Dafny SDK Interface] --> ToNativeW[Convert input \nto native]
+    ToDafnyW[Convert output \nto Dafny] --> WrappingInterface
+end
+
+subgraph Library[Native Library]
+    direction RL
+    Interface[Public Interface] --> ToDafny[Convert input \nto Dafny]
+    ToDafny --> DafnyImplementation[Dafny implementation]
+    DafnyImplementation --> ToNative[Convert output \nto native]
+    ToNative --> Interface
+end
+
+ToNativeW --> Interface
+Interface --> ToDafnyW
+
+DafnyTest --> WrappingInterface
+WrappingInterface --> DafnyTest
+
+```
+
+The goal is for Dafny developers
+to approximate native calls into Dafny libraries.
+`smithy-dafny` uses the wrapped library feature
+to test the correctness of polymorph-runtime features.
+
+Like this: `var output := ToDafny(ToNative(Work(ToDafny(ToNative(input)))))`
+
+## Polymorph development
+
+When developing `polymorph-X`
+the first thing to ask is:
+Does `smithy-X` exist?
+
+If it does exist then:
+`polymorph-X` SHOULD re-use as much `smithy-X` code-gen as possible.
+`polymorph-X` MUST use the same native types as `smithy-X`.
+
+If it does not exists then look at the AWS SDK in X:
+`polymorph-X` SHOULD use the same native types as the native AWS SDK in X runtime.
+
+Regardless `polymorph-X` MUST convert these native types
+to the Dafny types produced by `smithy-dafny`.
+
+```mermaid
+---
+title: Generated Code
+---
+flowchart LR
+
+subgraph Project
+    NativeInterface
+    NativeTypes
+    ToNative
+    ToDafny
+    DafnyTypes
+    DafnyInterface
+end
+
+subgraph Polymorph
+    SmithyNative
+end
+
+Model --> Polymorph
+Polymorph --> NativeInterface
+SmithyNative --> NativeTypes
+Polymorph --> ToNative
+Polymorph --> ToDafny
+Model --> Dafny
+Dafny --> DafnyTypes
+Dafny --> DafnyInterface
+```
+
+In this way the dual inputs and output
+of the type conversion dual functions are defined.
+
+```mermaid
+---
+title: Type conversion dual
+---
+flowchart LR
+NativeTypes --> ToDafny --> DafnyTypes
+DafnyTypes --> ToNative --> NativeTypes
+```
+
+However there may be many to one relationships.
+Perhaps `string` and `char[]`.
+These two would convert to `ValidUTF8Bytes` or `seq<uint8>`.
```

Comment:
I'd pick an example language and be very precise. "For example in Java, either a `java.lang.String` or a `char[]` is a common way of providing string data. Either one should be converted to the same idiomatic representation of string in Dafny (either `ValidUTF8Bytes` or `string` or...)."

That's not actually the best example since `java.lang.String` is by far the most common. Rust might be better since it has three different flavors of strings. Also strings are an annoying case for Dafny right here since there are multiple choices, so you're actually jumping to "many to many" right away.

Perhaps blobs are even better in general: interfaces very commonly accept either a ByteBuffer or a byte[] + start + length triple in Java.